### PR TITLE
New package to auto-generate public API documentation

### PR DESCRIPTION
**Pull request by oandregal**
This PR introduces a new package called `docgen` to help us generate the _public API_ of our code. Given an entry point file, it outputs the existing ES6 export statements and their corresponding JSDoc comments in human-readable format.

## Usage

> npx docgen <relative-path-to-entry-point.js>

This command will generate a file named `entry-point-api.md` containing all the exports and its JSDoc comments.

**Example:**

Let's use it with the `wordcount` package:

> npx docgen packages/wordcount/src/index.js

and then inspect the output

> less packages/wordcount/src/index-api.md

Find more details, documentation, and examples in the [README.md](https://github.com/WordPress/gutenberg/blob/try/api-docs/packages/docgen/README.md)

## Testing

Play with this to create documentation for a Gutenberg package, components, etc and report back!

## Coda

Note that this is only intended to work with native ES6 modules, not CommonJS. We have some packages used as config files or hooking into other tools (webpack, babel, etc) that are unsupported. See [coverage.md](packages/docgen/coverage.md).

### CODE REVIEW

**Comment by chrisvanpatten** [comment]
It would be helpful to clarify how this relates to the work that @herregroen has done in getting jsdoc parsing for the core DevHub :)

**Comment by herregroen** [comment]
From what I can see doctrine is used here to parse the JSDoc. It seems to have a format that's at least similar to what JSDoc uses itself.

I've written https://github.com/herregroen/jsdoc-parser which functions as a JSDoc template and outputs to a json format that can be imported by https://github.com/WordPress/phpdoc-parser.

Ideally this would have an option to specify a JSON output format that would also produce docs that can be imported by the above plugin.

If that's the case then once https://meta.trac.wordpress.org/ticket/3063 gets resolved this could be used to generate the docs to be imported on developer.wordpress.org.

**Comment by mcsf** [comment]
> From what I can see doctrine is used here to parse the JSDoc. It seems to have a format that's at least similar to what JSDoc uses itself.
> 
> I've written https://github.com/herregroen/jsdoc-parser which functions as […]

@nosolosw and @herregroen, seems like we run the risk of duplicating efforts here. The way I see it, the atoms we need are:
- Extracting data from Gutenberg JS code. In particular, extract function signatures and JSDoc information.
- Produce this data in a language-agnostic format. Ideally, one that is already supported by what we use to generate pages on developer.wordpress.org from PHP code.
- Handle this kind of data more or less regardless of the language it refers to (JS or PHP).
- Run it through any kind of formatter (e.g., Markdown; some specific HTML template).

It appears we have loose pieces of work for roughly all of these, but they are spread out across properties and tickets and it's hard to keep track of them. Examples I've found:
- WordPress/phpdoc-parser#207: Add programming language support
- [meta#3063](https://meta.trac.wordpress.org/ticket/3063): Show JavaScript documentation on developer.wordpress.org
- [herregroen/jsdoc-parser](https://github.com/herregroen/jsdoc-parser/): Proposed JSDoc template
- WordPress/gutenberg#13329 (this very PR): docgen: JSDoc parser, Markdown formatter, etc.

Can we sync up on our needs and figure out how to converge our work? Let's make sure that the atoms can be worked on completely in parallel, which naturally entails agreeing on the dialects involved—to represent parsed data, etc.

**Comment by chrisvanpatten** [comment]
@nosolosw SUPER excited to see us consolidating efforts here! Love it!

**Comment by mcsf** [comment]
Great to know! Let us know once you'd like a review.

**Comment by aduth** [comment]
A couple high-level thoughts:

- Doctrine was announced officially to be end-of-life ([reference](https://eslint.org/blog/2018/11/jsdoc-end-of-life))
- From your "TODO" notes of your `coverage.md` file, we should definitely not document `__unstable` APIs. I'd suggest even skipping `__experimental` APIs ([reference](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#experimental-apis)). Further, I think we might also take hints ([`@deprecated`](http://usejsdoc.org/tags-deprecated.html) JSDoc tag) to avoid documenting deprecated APIs, or otherwise clearly mark them as deprecated in the generated result.

**Comment by nosolosw** [comment]
> Doctrine was announced officially to be end-of-life 

Ah, gotcha, thanks for the heads up @aduth. Will take look at alternatives.

> From your "TODO" notes of your `coverage.md` file, we should definitely not document `__unstable` APIs. I'd suggest even skipping `__experimental` APIs ([reference](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#experimental-apis)). Further, I think we might also take hints ([`@deprecated`](http://usejsdoc.org/tags-deprecated.html) JSDoc tag) to avoid documenting deprecated APIs, or otherwise clearly mark them as deprecated in the generated result.

Thanks for catching the `@deprecated` tag. In that case, I suggest adding a notice (for example, along the lines of [`UNSAFE_compomentWillMount`](https://reactjs.org/docs/react-component.html#unsafe_componentwillmount)).

For `__unstable`/`__experimental`/etc I was thinking about moving that decission to userland. My current thinking is to provide a way to tell the doc generator to omit certain symbol names.

File: packages/docgen/README.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+Install the module
+
+```bash
+npm install @wordpress/docgen --save-dev
+```
+
+## Usage
+
+`node src/cli.js <entry-point.js>`
```
Comment (gziolo) [inline packages/docgen/README.md]: can it use bash syntax for consistency?

File: packages/docgen/README.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+* **--debug**: Run in debug mode, which outputs some intermediate files useful for debugging
+* **--append** `(String)`: Title of the Markdown section to append the generated documentation to. It takes precedence over and bypasses the *--formatter* option.
+
+## Examples
+
+### Default export
+
+Entry point:
+
+```js
```
Comment (gziolo) [inline packages/docgen/README.md]: Out of curiosity, do you use the tool to generate the output of those examples?
If not, maybe it would make sense to find a way to do it, format can change so it would be easier to keep it in sync :)

File: packages/docgen/package.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		"mdast-util-inject": "1.1.0",
+		"remark": "9.0.0",
+		"remark-parse": "6.0.3",
+		"tape": "4.9.2",
+		"unified": "7.1.0"
+	},
+	"publishConfig": {
+		"access": "public"
+	},
+	"scripts": {
```
Comment (gziolo) [inline packages/docgen/package.json]: This is different than all other packages. Unit tests are executed from the root folder for everything else. It should be documented for future maintainers how to run those tests. You also need to keep in mind that to install `devDependencies` in here someone will have to explicitly go to this folder. It won't get installed with `npm i` in the root folder. Only `dependencies` get hoisted.

File: packages/docgen/package.json [hunk 4]
```diff
@@ -0,0 +1,44 @@
+{
+	"name": "docgen",
+	"version": "1.0.0-beta.0",
+	"description": "Autogenerate public API documentation from exports and JSDoc comments.",
+	"author": "The WordPress Contributors",
+	"license": "GPL-2.0-or-later",
+	"keywords": [
+		"jsdoc",
```
Comment (gziolo) [inline packages/docgen/package.json]: We always put `wordpress` keyword in this array.

File: packages/docgen/README.md (hunk 2 above)
Comment (nosolosw) [reply packages/docgen/README.md]: Yeah, I manually created some examples and then generated this output. I agree a more streamlined process would be great.

Perhaps a potential way to improve this would be what you proposed at https://github.com/WordPress/gutenberg/pull/13856#discussion_r256843355 If we had a way to mark a Markdown file with tokens (instead of the current system that uses a section title) it'd be useful for this as well.

File: packages/docgen/package.json (hunk 4 above)
Comment (nosolosw) [reply packages/docgen/package.json]: b08b8d3cf3824addaac12c546316c909e7a38194

File: packages/docgen/README.md [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+Install the module
+
+```bash
+npm install @wordpress/docgen --save-dev
+```
+
+## Usage
+
+```bash
```
Comment (nosolosw) [inline packages/docgen/README.md]: Rebasing from master seems to have deleted your previous comment, @gziolo. For reference 399629ad2ae8b067aa85f7174d77a8184fa4c9f7 added `bash` syntax to this as well.

File: packages/docgen/package.json (hunk 3 above)
Comment (nosolosw) [reply packages/docgen/package.json]: Happy to adapt this to the way other packages are set up. I'm just unfamiliar with the lerna monorepo setup and haven't found any doc about how should we create new packages. Do you have any reference I can check? Or specific steps?

Comment (gziolo) [reply packages/docgen/package.json]: Technically speaking we can leave it as is. I just wanted to make sure you are aware of the implications of such setup.

For reference, this is how you can run a script from the package maintained by Lerna:
https://github.com/WordPress/gutenberg/blob/master/package.json#L154
`lerna run build` in particular is the way to run all `build` scripts in packages. See more at https://github.com/lerna/lerna/tree/master/commands/run

**Comment by nosolosw** [comment]
Did some research on doctrine alternatives (will update if needed):

Library | Latest publish (months ago) | npm weekly downloads (thousands) | Comments
--- | --- | --- | --- 
eslint-plugin-jsdoc | 1 | [30](https://www.npmjs.com/package/eslint-plugin-jsdoc) | Uses comment-parser and jsdoctypeparser.
comment-parser | 1 | [120](https://www.npmjs.com/package/comment-parser) | Parses JSDoc comment, but not types.
jsdoctypeparser | 36 | [100](https://www.npmjs.com/package/jsdoctypeparser) | Google Closure and JSDoc type parser.
jsdoc | 12 | [235](https://www.npmjs.com/package/jsdoc) | Uses catharsis for type parser and their own thing for the rest.
catharsis | 24 | [224](https://www.npmjs.com/package/catharsis) | Google Closure and JSDoc type parser.
doctrine | 3 | [8.000](https://www.npmjs.com/package/doctrine) | Developed their own parsers.

**Comment by nosolosw** [comment]
At this point, I think it makes sense to keep `doctrine` until a better alternative is developed.

I don't see a mature alternative to `doctrine` yet: I haven't found how to use `jsdoc` as a library to parse a JSDoc comment, and other libraries such as `comment-parser` have some gotchas (such as: it isn't able to [parse tags without a name out of the box](https://github.com/yavorskiy/comment-parser/issues/21) -`return`, `link`, `see`, `deprecated`- although it is fixable by writing custom parser for each tag.

In 5b3fcb7, b8ba3d0, 1bde19e, and 1bdcb95 I've added some tests and wrapped everything related to `doctrine` within the [`get-jsdoc-from-token`](https://github.com/WordPress/gutenberg/blob/1bdcb9592b98921128c82f6e2a97716f678b6878/packages/docgen/src/get-jsdoc-from-token.js) module, so it's easy to switch without affecting the `docgen` public API when time comes.

File: packages/docgen/README.md (hunk 1 above)
Comment (nosolosw) [reply packages/docgen/README.md]: 399629a

File: packages/docgen/package.json (hunk 3 above)
Comment (nosolosw) [reply packages/docgen/package.json]: Fixed what you mentioned:

- tests are now migrated to jest, so they're picked by it and executed as any other WordPress package.
- dependencies are now hoisted.

File: packages/docgen/README.md (hunk 2 above)
Comment (nosolosw) [reply packages/docgen/README.md]: I've implemented a new CLI option `--to-token` that allows inserting the autogenerated docs within the Markdown file.

**Comment by nosolosw** [comment]
I've addressed Greg comments and this is ready for a second round of review.

cc @herregroen @mcsf @chrisvanpatten @aduth and any other person interested :)

File: packages/docgen/README.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+npx docgen <entry-point.js>
+```
+
+This command will generate a file named `entry-point-api.md` containing all the exports and its JSDoc comments.
```
Comment (chrisvanpatten) [inline packages/docgen/README.md]:
```suggestion
This command will generate a file named `entry-point-api.md` containing all the exports and their JSDoc comments.
```

File: packages/docgen/README.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+  * *docPath* `(String)`: path of the output document to generate.
+  * *symbols* `(Array)`: the symbols found.
+* **--ignore** `(RegExp)`: A regular expression used to ignore symbols whose name match it.
+* **--output** `(String)`: Output file that will contain the API documentation.
+* **--to-section** `(String)`: Append generated documentation to this section in the Markdown output. To be used by the default Markdown formatter. Depends on `--output` and bypasses the custom `--formatter` passed, if any.
+* **--to-token**: Embed generated documentation within the start and end tokens in the Markdown output. To be used by the default Markdown formatter.Depends on `--output` and bypasses the custom `--formatter` passed, if any.
+  * Start token: `<!-- START TOKEN(Autogenerated API docs) -->`
+  * End token: `<!-- END TOKEN(Autogenerated API docs) -->`
+* **--use-token** `(String)`: This options allows to customize the string between the tokens. For example, `--use-token my-api` will look up for the start token `<!-- START TOKEN(my-api) -->` and the end token `<!-- END TOKEN(my-api) -->`. Depends on `--to-token`.
```
Comment (chrisvanpatten) [inline packages/docgen/README.md]:
```suggestion
* **--use-token** `(String)`: This options allows you to customize the string between the tokens. For example, `--use-token my-api` will look up for the start token `<!-- START TOKEN(my-api) -->` and the end token `<!-- END TOKEN(my-api) -->`. Depends on `--to-token`.
```

File: packages/docgen/README.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+* **--debug**: Run in debug mode, which outputs some intermediate files useful for debugging
```
Comment (chrisvanpatten) [inline packages/docgen/README.md]:
```suggestion
* **--debug**: Run in debug mode, which outputs some intermediate files useful for debugging.
```

File: packages/docgen/README.md [hunk 9]
```diff
@@ -0,0 +1,250 @@
+# `docgen`
+
+`docgen` helps you to generate the _public API_ of your code. Given an entry point file, it outputs the ES6 export statements and their corresponding JSDoc comments in human-readable format.
+
+Some characteristics:
+
+* If the export statement doesn't contain any JSDoc, it'll look up for JSDoc up to the declaration.
```
Comment (chrisvanpatten) [inline packages/docgen/README.md]: I'm not sure I understand this. Can you clarify?

Comment (chrisvanpatten) [reply packages/docgen/README.md]: I guess you're saying if there's no JSDoc on the export, it will look for the JSDoc with the exported thing itself?

**Review by chrisvanpatten — DISMISSED** [review]

**Review by mkaz — APPROVED** [review]
Great work! 👍 This works nicely as described. I tested it on a few of the existing ES6 packages and it worked as expected. The next piece will be how to hook and generate for existing documentation but that will be another PR.

One small suggestion, but wouldn't consider it a blocker since you do have the example command above, is adding in the examples the file names and command used, for example:

Entry point: `index.js`

```
/**
 * Adds two numbers.
 *
 * @param {number} term1 First number.
 * @param {number} term2 Second number.
 * @return {number} The result of adding the two numbers.
 */
export default function addition( term1, term2 ) {
	// Implementation would go here.
}
```
Command: `npx docgen index.js`

Output: `index-api.md`

```
# API

## default

[example.js#L8-L10](example.js#L8-L10)

Adds two numbers.

**Parameters**

- **term1** `number`: First number.
- **term2** `number`: Second number.

**Returns**

`number` The result of adding the two numbers.
```

File: packages/docgen/README.md (hunk 9 above)
Comment (nosolosw) [reply packages/docgen/README.md]: Exactly. For example:

```js
/**
 * Adds two numbers.
 *
 * @param {number} term1 First number.
 * @param {number} term2 Second number.
 * @return {number} The result of adding the two numbers.
 */
function addition( term1, term2 ) {
	// Implementation would go here.
}

export default addition;
```

The export statement doesn't have any JSDoc, but the symbol declaration does.

**Comment by nosolosw** [comment]
> The next piece will be how to hook and generate for existing documentation but that will be another PR.

I was tempted to do it here as well, but it'd add unnecessary noise. I actually added the command for the `e2e-test-utils` package, though, just because it was the first testbed for this PR and I wanted to demonstrate how it would fit.

> One small suggestion, but wouldn't consider it a blocker since you do have the example command above, is adding in the examples the file names and command used

Nice suggestion. Implemented at 0f364a2.

**Comment by nosolosw** [comment]
This was merged everyone, let's get the docs :tada: started!

File: packages/docgen/src/engine.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
+/**
```
Comment (aduth) [inline packages/docgen/src/engine.js]: I wonder if we should have avoided placing these files in a directory called `src/`, because for every other package we use the `src` directory to indicate pre-transpiled code, with the expectation there'd be a complementing `build` directory as output. I encountered this today when auditing our transpiled packages for `@babel/runtime` and became confused when this package was not respecting the convention.

In other words, I see no reason why all of these files couldn't just live in the root of the package, or maybe we could create a different convention for organizing files within packages which don't need transpiling.

Comment (aduth) [reply packages/docgen/src/engine.js]: In fact, `src/` is used as a trigger to process files via Babel.

You can see that `docgen` includes `build` and `build-module` outputs, even though they're effectively dead code:

https://unpkg.com/@wordpress/docgen@1.0.0/

Related code:

https://github.com/WordPress/gutenberg/blob/9c3dbde9f7585ed776fa52a3720ad581d556f865/bin/packages/build.js#L193-L200

Comment (gziolo) [reply packages/docgen/src/engine.js]: We can change this logic and decide whether a given package should be transpiled based on the existance of `module` entry in `package.json` file.

Comment (gziolo) [reply packages/docgen/src/engine.js]: I see what you mean. `docgen` never uses transpiled code but it's still published to npm.

Comment (nosolosw) [reply packages/docgen/src/engine.js]: > In fact, src/ is used as a trigger to process files via Babel.

Didn't know that until now.

> We can change this logic and decide whether a given package should be transpiled based on the existance of module entry in package.json file.

I like this. It allows us to store code within `src/` for complex packages like `docgen` that require multiple files. It's also in line with the idea we discussed somewhere about favoring `package.json` config over maintaining ignore lists.

File: packages/docgen/src/index.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+	return result.ir || undefined;
+};
+
+const processFile = ( rootDir, inputFile ) => {
+	try {
+		const data = fs.readFileSync( inputFile, 'utf8' );
+		currentFileStack.push( inputFile );
+		const relativePath = path.relative( rootDir, inputFile );
+		const result = engine( relativePath, data, getIRFromRelativePath( rootDir, last( currentFileStack ) ) );
+		currentFileStack.pop( inputFile );
```
Comment (aduth) [inline packages/docgen/src/index.js]: `Array#pop` takes no arguments. It could be safely removed.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/pop

File: packages/docgen/src/get-export-entries.js [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+		const getLocalName = ( t ) => {
+			let name;
+			switch ( t.declaration.type ) {
+				case 'Identifier':
+					name = t.declaration.name;
+					break;
+				case 'AssignmentExpression':
+					name = t.declaration.left.name;
+					break;
+				//case 'FunctionDeclaration'
```
Comment (aduth) [inline packages/docgen/src/get-export-entries.js]: Should have been removed?

File: packages/docgen/src/test/fixtures/default-class-named/exports.json [hunk 13]
```diff
@@ -0,0 +1,101 @@
+{
```
Comment (aduth) [inline packages/docgen/src/test/fixtures/default-class-named/exports.json]: How are these files generated?

Comment (nosolosw) [reply packages/docgen/src/test/fixtures/default-class-named/exports.json]: Using docgen's `--debug` [mode](https://github.com/WordPress/gutenberg/blob/master/packages/docgen/README.md#cli-options). 

From the root of the repo:

>npx docgen --debug packages/docgen/src/test/fixtures/default-class-named/code.js

Comment (nosolosw) [reply packages/docgen/src/test/fixtures/default-class-named/exports.json]: I seem to have tweaked the exports files removing the enclosing `[]`. In retrospect, don't know why I thought that was a good idea.

File: packages/docgen/src/get-export-entries.js (hunk 12 above)
Comment (nosolosw) [reply packages/docgen/src/get-export-entries.js]: From memory, I think it documents the cases that fall-through.

File: packages/docgen/src/index.js (hunk 11 above)
Comment (nosolosw) [reply packages/docgen/src/index.js]: Mind to add this in the `@private` PR?

File: packages/docgen/src/get-export-entries.js (hunk 12 above)
Comment (aduth) [reply packages/docgen/src/get-export-entries.js]: > From memory, I think it documents the cases that fall-through.

It could also be explicit, no?

```js
switch ( t.declaration.type ) {
	// ...

	case 'FunctionDeclaration':
	case 'ClassDeclaration':
	default:
		// ...
}
```

_(If it were a web bundle, we may have concern about the "unnecessary" bytes added, but it's not in this instance)_

I'd just worry that this might not be obvious to the future maintainer without at least a hint (inline comment) to explain the intention.

File: packages/docgen/src/index.js (hunk 11 above)
Comment (aduth) [reply packages/docgen/src/index.js]: > Mind to add this in the `@private` PR?

I made a small pull request: #15397
